cartographer_ros/cartographer_ros/node.cc

```diff
@@ -206,69 +206,57 @@ void Node::PublishSubmapList(const ::ros::WallTimerEvent& unused_timer_event) {
-    stamped_transform.header.stamp = ros::Time::now();
-  }
-
-  const auto published_to_tracking =
-      map_builder_bridge_.sensor_bridge(trajectory_id_)
-          ->tf_bridge()
-          .LookupToTracking(last_pose_estimate.time, options_.published_frame);
-  if (published_to_tracking != nullptr) {
-    if (options_.provide_odom_frame) {
-      std::vector<geometry_msgs::TransformStamped> stamped_transforms;
-
-      stamped_transform.header.frame_id = options_.map_frame;
-      stamped_transform.child_frame_id = options_.odom_frame;
-      stamped_transform.transform = ToGeometryMsgTransform(local_to_map);
-      stamped_transforms.push_back(stamped_transform);
-
-      stamped_transform.header.frame_id = options_.odom_frame;
-      stamped_transform.child_frame_id = options_.published_frame;
-      stamped_transform.transform =
-          ToGeometryMsgTransform(tracking_to_local * (*published_to_tracking));
-      stamped_transforms.push_back(stamped_transform);
-
-      tf_broadcaster_.sendTransform(stamped_transforms);
+  for (const auto& entry : map_builder_bridge_.GetTrajectoryStates()) {
+    const auto& trajectory_state = entry.second;
+
+    geometry_msgs::TransformStamped stamped_transform;
+    stamped_transform.header.stamp = ToRos(trajectory_state.pose_estimate.time);
+
+    const auto& tracking_to_local = trajectory_state.pose_estimate.pose;
+    const Rigid3d tracking_to_map =
+        trajectory_state.local_to_map * tracking_to_local;
+
+    // We only publish a point cloud if it has changed. It is not needed at high
+    // frequency, and republishing it would be computationally wasteful.
+    if (trajectory_state.pose_estimate.time != last_scan_matched_point_cloud_time_) {
+      scan_matched_point_cloud_publisher_.publish(ToPointCloud2Message(
+          carto::common::ToUniversal(trajectory_state.pose_estimate.time),
+          options_.tracking_frame,
+          carto::sensor::TransformPointCloud(
+              trajectory_state.pose_estimate.point_cloud,
+              tracking_to_local.inverse().cast<float>())));
+      last_scan_matched_point_cloud_time_ = trajectory_state.pose_estimate.time;
     } else {
-      stamped_transform.header.frame_id = options_.map_frame;
-      stamped_transform.child_frame_id = options_.published_frame;
-      stamped_transform.transform =
-          ToGeometryMsgTransform(tracking_to_map * (*published_to_tracking));
-      tf_broadcaster_.sendTransform(stamped_transform);
+      // If we do not publish a new point cloud, we still allow time of the
+      // published poses to advance.
+      stamped_transform.header.stamp = ros::Time::now();
+    }
+
+    if (trajectory_state.published_to_tracking != nullptr) {
+      if (options_.provide_odom_frame) {
+        std::vector<geometry_msgs::TransformStamped> stamped_transforms;
+
+        stamped_transform.header.frame_id = options_.map_frame;
+        stamped_transform.child_frame_id = options_.odom_frame;
```

Comment:
Should we add a `TODO` that the `odom_frame` (and also `published_frame`?) must be per-trajectory when we fully support the multi-robot use case?